Core/src/RequestWrapper.php

```diff
@@ -305,48 +307,58 @@ private function applyHeaders(RequestInterface $request, array $options = [])
             unset($options['retryHeaders']);
         }
 
+        $request = Utils::modifyRequest($request, ['set_headers' => $headers]);
+
         if ($this->shouldSignRequest) {
             $quotaProject = $this->quotaProject;
-            $token = null;
 
             if ($this->accessToken) {
-                $token = $this->accessToken;
+                $request = $request->withHeader('authorization', 'Bearer ' . $this->accessToken);
             } else {
                 $credentialsFetcher = $this->getCredentialsFetcher();
-                $token = $this->fetchCredentials($credentialsFetcher)['access_token'];
+                $request = $this->addAuthHeaders($request, $credentialsFetcher);
 
                 if ($credentialsFetcher instanceof GetQuotaProjectInterface) {
                     $quotaProject = $credentialsFetcher->getQuotaProject();
                 }
             }
 
-            $headers['Authorization'] = 'Bearer ' . $token;
-
             if ($quotaProject) {
-                $headers['X-Goog-User-Project'] = [$quotaProject];
+                $request = $request->withHeader('X-Goog-User-Project', $quotaProject);
             }
         }
 
-        return Utils::modifyRequest($request, ['set_headers' => $headers]);
+        return $request;
     }
 
     /**
-     * Fetches credentials.
+     * Adds auth headers to the request.
      *
-     * @param FetchAuthTokenInterface $credentialsFetcher
+     * @param RequestInterface $request
+     * @param FetchAuthTokenInterface $fetcher
      * @return array
      * @throws ServiceException
      */
-    private function fetchCredentials(FetchAuthTokenInterface $credentialsFetcher)
+    private function addAuthHeaders(RequestInterface $request, FetchAuthTokenInterface $fetcher)
     {
         $backoff = new ExponentialBackoff($this->retries, $this->getRetryFunction());
 
         try {
             return $backoff->execute(
-                function () use ($credentialsFetcher) {
-                    if ($token = $credentialsFetcher->fetchAuthToken($this->authHttpHandler)) {
-                        return $token;
+                function () use ($request, $fetcher) {
+                    if (!$fetcher instanceof UpdateMetadataInterface ||
+                         ($fetcher instanceof FetchAuthTokenCache &&
+                            !$fetcher->getFetcher() instanceof UpdateMetadataInterface
+                         )
+                    ) {
+                        if ($token = $fetcher->fetchAuthToken()) {
```

Comment:
Let's add a helpful comment here (since this logic is not super easy to read and understand)
```suggestion
                    ) {
                        // This covers an edge case where the token fetcher does not implement UpdateMetadataInterface,
                        // which only would happen if a user implemented a custom fetcher 
                        if ($token = $fetcher->fetchAuthToken()) {
```